dev/bots/utils.dart

```diff
@@ -2,87 +2,210 @@
+  }
+  print('$red╚${"═" * width}');
+  // Normally, "print" actually prints to the log. To make the errors visible,
+  // and to include useful context, we now print the entire log up to this
+  // point, and clear it. Subsequent messages will continue to not be logged
+  // until there is another error.
+  _pendingLogs.forEach(_printLoudly);
+  _pendingLogs.clear();
   _errorMessages.addAll(messages);
   _hasError = true;
+  if (onError != null) {
+    onError!();
+  }
 }
 
 @visibleForTesting
 void resetErrorStatus() {
   _hasError = false;
   _errorMessages.clear();
+  _pendingLogs.clear();
+  _hideTimer?.cancel();
+  _hideTimer = null;
 }
 
 Never reportErrorsAndExit() {
+  _hideTimer?.cancel();
+  _hideTimer = null;
   print(redLine);
   print('For your convenience, the error messages reported above are repeated here:');
   _errorMessages.forEach(print);
   print(redLine);
   system.exit(1);
 }
 
-String get clock {
-  final DateTime now = DateTime.now();
-  return '$reverse▌'
-         '${now.hour.toString().padLeft(2, "0")}:'
-         '${now.minute.toString().padLeft(2, "0")}:'
-         '${now.second.toString().padLeft(2, "0")}'
-         '▐$reset';
+void printProgress(String message) {
+  _pendingLogs.clear();
+  _hideTimer?.cancel();
+  _hideTimer = null;
+  print('$clock $message $reset');
+  if (hasColor) {
+    // We set up a timer to switch to verbose mode when the tests take too long,
+    // so that if a test hangs we can see the logs.
+    // (This is only supported with a color terminal. When the terminal doesn't
+    // support colors, we just print everything verbosely, that way in CI there's
+    // nothing hidden.)
+    _hideTimer = Timer(_quietTimeout, () {
+      _hideTimer = null;
+      _pendingLogs.forEach(_printLoudly);
+      _pendingLogs.clear();
+    });
+  }
 }
 
-String prettyPrintDuration(Duration duration) {
-  String result = '';
-  final int minutes = duration.inMinutes;
-  if (minutes > 0) {
-    result += '${minutes}min ';
+final Pattern _lineBreak = RegExp(r'[\r\n]');
+
+void _printQuietly(Object? message) {
+  // The point of this function is to avoid printing its output unless the timer
+  // has gone off in which case we assume we're in verbose mode and print
+  // everything. To show that progress is still happening though, we can't just
+  // show nothing, so instead we show the last line of output and keep
+  // overwriting it. To do this in color mode we have to carefully measure the
+  // line of text ignoring color codes, which is what the parser below does.
+  if (_hideTimer != null) {
+    _pendingLogs.add(message.toString());
+    String line = '$message'.trimRight();
+    final int start = line.lastIndexOf(_lineBreak) + 1;
+    int index = start;
+    int length = 0;
+    while (index < line.length && length < stdout.terminalColumns) {
+      if (line.codeUnitAt(index) == kESC) {
+        index += 1;
+        if (index < line.length && line.codeUnitAt(index) == 0x5B) { // [
```

Comment:
Can you also convert this and below to const as well?

Reply:
yup